libservo: Group site data entries by eTLD+1 and simplify related tests

janvarga opened the pull request:
This PR updates `SiteDataManager::site_data` to group entries by eTLD+1
instead of origin, matching how site data is typically presented in browser
UI and aligning the API with user expectations.

The earlier patches in the series clean up the related tests (deterministic
server ordering, clearer expectations, sorted output) to make the change straightforward.

This also lays groundwork for future extensions such as cookie related site
data reporting.

Testing: The integration test was updated to reflect the new eTLD+1 grouping.

janvarga committed:
libservo: Use realistic hostnames instead of localhost in site data tests

Replace `localhost` with realistic hostnames in the site data tests. This
better reflects real world usage and is required for the upcoming grouping
by eTLD+1.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Group site data entries by eTLD+1 instead of origin

Change `SiteDataManager::site_data` to group storage entries by eTLD+1 rather
than by origin. This aligns the API with user expectations and with how
browsers typically present site data in UI, where related origins belonging to
the same site are shown together.

This also prepares the API for adding cookie support, which would be
problematic to present at the origin level.

Both public and private storage are still included, and the returned list
remains sorted by site name.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Move rustdoc to the correct location

The rustdoc comment for `SiteDataManager` was previously attached to the
wrong item. Move it so that it documents the struct where it belongs.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Sort sites by name in `SiteDataManager::site_data`

Make `site_data` return sites sorted by site name. Embedders usually display
sites in sorted order anyway, and this also has the nice side effect of
simplifying the tests.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Use explicit site names in assertions for clarity

Replace urls in the assertions with explicit site names. This makes the
expected results clearer in the test.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Create test HTTP servers up front in deterministic order

Create the lightweight HTTP servers in advance and sort them so that the
first server is consistently first in alphabetical order. This prepares the
test for follow-up patches that further simplify the checks.

With this change the servers can also be closed incrementally when they are
no longer needed (previously they had to be kept alive to avoid reusing the
same port).

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
net: Derive Debug for Server to support Vec::try_into usage

Add a `Debug` implementation for `Server` so that conversions such as
`Vec::try_into()` compiles.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Replace tuple expectations with explicit `SiteData` values

Stop constructing expected values from tuples and build them explicitly
instead. This prepares the ground for removing the `sites_equal_unordered`
helper in a follow-up patch.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga commented:
https://github.com/janvarga/servo/actions/runs/20506580385

yezhizhen commented on components/servo/site_data_manager.rs:
```diff
@@ -135,10 +136,14 @@ impl SiteDataManager {
-            .collect()
+            .collect();
+
+        result.sort_by_key(|a| a.name());
```
```suggestion
        result.sort_by_key(SiteData::name);
```

janvarga replied:
Ok

yezhizhen commented on components/servo/site_data_manager.rs:
```diff
@@ -91,21 +90,23 @@ impl SiteDataManager {
         let mut add_origins = |origins: Vec<OriginDescriptor>, storage_type: StorageType| {
             for origin in origins {
+                let url = ServoUrl::parse(&origin.name).unwrap();
+                let domain = registered_domain_name(&url).unwrap().to_string();
```
Ditto.

yezhizhen commented on components/servo/site_data_manager.rs:
```diff
@@ -28,31 +30,16 @@ bitflags! {
 
 impl SiteData {
-    pub fn new(name: String, storage_types: StorageType) -> SiteData {
+    pub fn new<N: Into<String>>(name: N, storage_types: StorageType) -> SiteData {
```
```suggestion
    pub fn new(name: impl Into<String>, storage_types: StorageType) -> SiteData {
```

Let's avoid generic here.

janvarga replied:
Yeah, that's better.

yezhizhen commented on components/servo/tests/site_data_manager.rs:
```diff
@@ -50,45 +35,67 @@ fn test_site_data() {
+        servers.push((server, url));
+    }
+
+    servers.sort_by(|(_, a), (_, b)| a.cmp(b));
```
```suggestion
    servers.sort_by_key(|(_, url)| url);
```

yezhizhen replied:
You are right! Let's revert this then.

yezhizhen reviewed (approved):
Just a couple of nits.

janvarga committed:
libservo: Address review feedback for site data grouping

This patch incorporates review feedback by applying the following cleanups
and refinements:

- Avoid generics in `SiteData::new`
- Replace `unwrap` with `expect` for URL parsing in `add_origins
- Gracefully handle failures when obtaining domain name in `add_origins`
- Refine use of `sort_by_key`

Signed-off-by: Jan Varga <jvarga@igalia.com>

yezhizhen commented on components/servo/site_data_manager.rs:
```diff
@@ -91,21 +90,23 @@ impl SiteDataManager {
 
         let mut add_origins = |origins: Vec<OriginDescriptor>, storage_type: StorageType| {
             for origin in origins {
+                let url = ServoUrl::parse(&origin.name).unwrap();
```
I was thinking about something like `.expect("<Why this should not fail>")`, instead of just `continue`.

janvarga replied:
Ok, added a warning.

yezhizhen commented:
Thanks!
